### PR TITLE
Add 'putResourceForURL' for pre-warming ambient cache

### PR DESCRIPTION
**Pull request by ChrisLoer**
This is an attempt at addressing issue https://github.com/mapbox/mapbox-gl-native/issues/13022 by directly exposing the `DefaultFileSource::put` method to the iOS/Android Offline functionality. This PR borrows heavily from @halset's https://github.com/mapbox/mapbox-gl-native/pull/12217, but is smaller in scope in that it's not trying to do bulk insertions (which can now be accomplished with the "mergeOfflineRegion" functionality).

This is a kind of awkward fit with `MGLOfflineStorage` and `OfflineManager` because at least the way we present the interface they're not directly related to the ambient cache (even though under the hood they're managing the same database). But it also felt like overkill to add a whole set of extra "ambient cache" wrapper classes.

@tobrun, I haven't figured out a good way to write tests for this on the Android side -- because we don't expose FileSource requests through to the Java code, there's no easy way to verify that after calling "put" the data round trips through the cache. Do you have any suggestions? The one thing I could think of is to write an explicit test function that does the verification on the C++ side instead of the Java side, but that feels pretty clunky.

/cc @ansis @kkaefer

### CODE REVIEW

**Comment by tobrun** [comment]
> @tobrun, I haven't figured out a good way to write tests for this on the Android side -- because we don't expose FileSource requests through to the Java code, there's no easy way to verify that after calling "put" the data round trips through the cache. Do you have any suggestions? The one thing I could think of is to write an explicit test function that does the verification on the C++ side instead of the Java side, but that feels pretty clunky.

It seems we are missing some APIs in comparison to iOS, I would love to bring them more together. Atm there isn't any easy way to verify this with a test from the java end, afaik only with a manual verification by inspecting the DB after the put call. To not block this PR from merging, I can look into doing this manual test? ticket out the missing API for FileSource requests on Android and the missing test for put?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineManager.java [hunk 1]
```diff
@@ -462,4 +462,25 @@ private native void createOfflineRegion(FileSource fileSource, OfflineRegionDefi
...
+   *
+   * @param url The URL of the resource to insert
+   * @param data Response data to store for this resource. Data is expected to be uncompressed;
+   *             internally, the cache will compress data as necessary.
+   * @param modified Optional "modified" response header, in seconds since 1970, or 0 if not set
+   * @param expires Optional "expires" response header, in seconds since 1970, or 0 if not set
+   * @param etag Optional "entity tag" response header
+   */
+  @Keep
+  public native void putResourceWithUrl(String url, byte[] data, long modified, long expires, String etag);
```
Comment (kkaefer) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineManager.java]: This API is missing support for `must-revalidate`

File: platform/darwin/src/MGLOfflineStorage.h [hunk 2]
```diff
@@ -339,6 +339,26 @@ MGL_EXPORT
...
+ * for in-progress requests, although subsequent requests should have
+ * access to the cached data.
+ *
+ * @param url The URL of the resource to insert
+ * @param data Response data to store for this resource. Data is expected to be uncompressed; internally, the cache will compress data as necessary.
+ * @param modified Optional "modified" response header
+ * @param expires Optional "expires" response header
+ * @param etag Optional "entity tag" response header
+ */
+-(void)putResourceWithUrl:(NSURL *)url data:(NSData *)data modified:(NSDate * _Nullable)modified expires:(NSDate * _Nullable)expires etag:(NSString * _Nullable)etag;
```
Comment (kkaefer) [inline platform/darwin/src/MGLOfflineStorage.h]: Same

**Comment by ChrisLoer** [comment]
> To not block this PR from merging, I can look into doing this manual test? ticket out the missing API for FileSource requests on Android and the missing test for put?

I ticketed the FileSource API at https://github.com/mapbox/mapbox-gl-native/issues/13126. Not sure if there's a strong use case for it outside of testing, but just testing might be enough to justify it.

For now I just did a basic manual test in the test app by modifying `SimpleMapActivity` to cache an arbitrary string at an arbitrary URL and then to call `setStyleURL` with that URL -- I verified that the bogus style loaded from the cache, failed to parse, triggered an invalidation of the element, and then a retry to hit the bogus URL (which 404ed).

**Pull request merged**